### PR TITLE
Added downsample and classweights

### PR DESCRIPTION
**Pull request by yinlou**
Added 'downsample' to config, downsample samples a portion of data given a label (stratified sampling).

Added 'classweights' to config. This is used for up-sampling by increasing
the weight on minority class.

Added unit test for downsampling.

### CODE REVIEW

**Comment by deerzq** [comment]
Unit test is failing: https://travis-ci.org/airbnb/aerosolve/builds/130624410#L1583

File: training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala [hunk 1]
```diff
@@ -255,7 +259,7 @@ object AdditiveModelTrainer {
     val expCorr = scala.math.exp(corr)
     val loss = scala.math.log(1.0 + 1.0 / expCorr)
     val grad = -label / (1.0 + expCorr)
-    val gradWithLearningRate = grad.toFloat * params.learningRate.toFloat
+    val gradWithLearningRate = grad.toFloat * params.learningRate.toFloat * classWeights(label.toInt)
```
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala]: additive model also has `regression` loss, the label for regression may not be integer. The customized sampling should  only be applied to classification case and if not specified not affecting the regression learning

Comment (yinlou) [inline training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala]: The classWeights only applies on classification so I only edited updateHinge and updateLogistic. Is 'hinge' also considered as regression?

Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala]: oh, never mind, then you may also want to add this to `updateLogistic`?

Comment (yinlou) [inline training/src/main/scala/com/airbnb/aerosolve/training/AdditiveModelTrainer.scala]: I think this line is under `updateLogistic` :), and you can find another line below for `updateHinge`

File: training/src/main/scala/com/airbnb/aerosolve/training/TrainingUtils.scala [hunk 2]
```diff
@@ -179,6 +203,16 @@ object TrainingUtils {
     }
   }
 
+  def getLabel(example: Example, loss: String, rankKey: String, threshold: Double): Int = {
+    val fv = example.getExample.get(0)
+    val label: Int = if (loss == "regression") {
+      TrainingUtils.getLabel(fv, rankKey).toInt
```
Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/TrainingUtils.scala]: why `toInt`?

Comment (yinlou) [inline training/src/main/scala/com/airbnb/aerosolve/training/TrainingUtils.scala]: Since we know this is a classification label, and I use `int` to create the maps (classweights and downsample), that will make the context clearer and can usually prevent potential bugs.

Comment (deerzq) [inline training/src/main/scala/com/airbnb/aerosolve/training/TrainingUtils.scala]: but the loss is regression, the label should be double. If this is returning classification label, perhaps we should not allow the loss to be 'regression'

Comment (yinlou) [inline training/src/main/scala/com/airbnb/aerosolve/training/TrainingUtils.scala]: Ah I see your point. Updated the code!

**Comment by deerzq** [comment]
Great work! LGTM after fixing build failure
